lib/handlers/pkg.put.js

```diff
@@ -93,54 +124,69 @@ const Parser = class Parser extends Duplex {
     }
 
     _write(chunk, enc, cb) {
-        this.parser.write(chunk);
+        this._parser.write(chunk);
         cb();
     }
 
     _final(cb) {
-        if (this._done) {
+        // Upload was successfull. Close the stream so the http response can is closed.
+        if (this._state >= STATE_SUCCESS) {
             this.push(null);
             cb();
-        } else {
-            this.once('done', () => {
+            return;
+        }
+
+        this.on('state', (state) => {
+            if (state >= STATE_SUCCESS) {
                 this.push(null);
                 cb();
-            });
-        }
+            }
+        });
     }
 }
 
-const handler = async (sink, req, org, name, version) => {
-    if (typeof org !== 'string' || org === '') {
-        throw new TypeError(
-            ':org is a required url parameter and must be a string'
-        );
-    }
+const init = (sink, incoming) => {
+    return new Promise((resolve, reject) => {
+        const parser = new Parser(incoming, sink);
 
-    if (typeof name !== 'string' || name === '') {
-        throw new TypeError(
-            ':name is a required url parameter and must be a string'
-        );
-    }
+        parser.on('state', (state) => {
+            if (parser._state === STATE_PROCESSING) {
+                resolve(outgoing);
+            }
+        });
+
+        // If incoming.request is handeled by pipeline, it will close
+        // to early for the http framework to handle it. Let the
+        // http framework handle closing incoming.request
+        incoming.request.pipe(parser);
 
-    if (typeof version !== 'string' || version === '') {
-        throw new TypeError(
-            ':version is a required url parameter and must be a string'
-        );
+        const outgoing = new HttpOutgoing();
+        outgoing.mimeType = 'application/octet-stream';
+
+        pipeline(parser, outgoing, (error) => {
+            if (error && parser._state === STATE_INITIALIZING) {
+                reject(error);
+                return;
+            }
```

Comment:
what happens if error but state is not STATE_INITIALIZING?

Reply:
Then one are passed the point where a http 4xx or 5xx can be sent. A http 200 is already sent and some data might have been pushed to the outgoing stream (ex; half a package is extracted and it then fails for some reason). What will happen then is that the stream is terminated. We are at a point where we can not send a error message to the user.

We should though log it server side.